File: src/memlore/domain/services/verification.py
from __future__ import annotations

from datetime import UTC, datetime

from memlore.domain.exceptions import ValidationError
from memlore.domain.models.audit_record import AuditRecord
from memlore.domain.models.enums import AuditAction, VerificationStatus
from memlore.domain.models.lore_entry import LoreEntry


def apply_verification(
    entry: LoreEntry,
    *,
    actor_id: str,
    now: datetime,
) -> tuple[LoreEntry, AuditRecord | None]:
    actor = actor_id.strip()
    if not actor:
        raise ValidationError("actor must be non-empty")

    if entry.verification_status == VerificationStatus.VERIFIED:
        return entry, None

    entry.verification_status = VerificationStatus.VERIFIED
    entry.verified_by = actor
    entry.verified_at = now
    entry.updated_at = now
    audit = AuditRecord(
        target_id=entry.id,
        action=AuditAction.VERIFY,
        actor_id=actor,
        created_at=now,
    )
    return entry, audit


def ensure_utc(value: datetime) -> datetime:
    if value.tzinfo is None:
        return value.replace(tzinfo=UTC)
    return value.astimezone(UTC)

File: internal/domain/verification.go
package domain

import (
	"strings"
	"time"
)

// ApplyVerification marks a lore entry verified on first call only.
// Returns the entry, an optional verify audit, and an error.
// Characterization: src/memlore/domain/services/verification.py
func ApplyVerification(entry LoreEntry, actorID string, now time.Time) (LoreEntry, *AuditRecord, error) {
	actor := strings.TrimSpace(actorID)
	if actor == "" {
		return LoreEntry{}, nil, validationError("actor must be non-empty")
	}

	if entry.VerificationStatus == VerificationInvalidated {
		return LoreEntry{}, nil, validationError("cannot verify an invalidated lore entry")
	}
	if IsSuperseded(entry) {
		return LoreEntry{}, nil, validationError("cannot verify a superseded lore entry")
	}

	if entry.VerificationStatus == VerificationVerified {
		return entry, nil, nil
	}

	now = EnsureUTC(now)
	entry.VerificationStatus = VerificationVerified
	entry.VerifiedBy = &actor
	entry.VerifiedAt = &now
	entry.UpdatedAt = now

	audit, err := NewAuditRecord(NewAuditRecordInput{
		TargetID:  entry.ID,
		Action:    AuditActionVerify,
		ActorID:   actor,
		CreatedAt: now,
	})
	if err != nil {
		return LoreEntry{}, nil, err
	}

	return entry, &audit, nil
}

// EnsureUTC normalizes a timestamp to UTC.
func EnsureUTC(value time.Time) time.Time {
	if value.IsZero() {
		return value
	}
	if value.Location() == time.UTC {
		return value
	}
	return value.UTC()
}
